core/vm/validate.go

```diff
@@ -0,0 +1,230 @@
+				return e(ErrTruncatedImmediate, i, "jump table truncated")
+			}
+			for j := 0; j < count; j++ {
+				if err := checkDest(code, &analysis, i+2+j*2, i+2*count+2, len(code)); err != nil {
+					return err
+				}
+			}
+			i += 1 + 2*count
+		case op == CALLF:
+			if i+2 >= len(code) {
+				return e(ErrTruncatedImmediate, i, "op=%s", op)
+			}
+			arg, _ := parseUint16(code[i+1:])
+			if arg >= len(metadata) {
+				return e(ErrInvalidSectionArgument, i, "arg %d, last section %d", arg, len(metadata))
+			}
+			i += 2
+		}
+		i += 1
+	}
+	// Code sections may not "fall through" and require proper termination.
+	// Therefore, the last instruction must be considered terminal.
+	if !jt[op].terminal {
+		return e(ErrInvalidCodeTermination, i, "ends with op %s", op)
+	}
+	if paths, err := validateControlFlow(code, section, metadata, jt); err != nil {
+		return err
+	} else if paths != count {
+		// TODO(matt): return actual position of unreacable code
+		return e(ErrUnreachableCode, 0, "")
+	}
+	return nil
+}
+
+// checkDest parses a relative offset at code[0:2] and checks if it is a valid jump destination.
+func checkDest(code []byte, analysis *bitvec, imm, from, length int) error {
+	if len(code) < imm+2 {
+		return io.ErrUnexpectedEOF
+	}
+	if analysis != nil && *analysis == nil {
+		*analysis = eofCodeBitmap(code)
+	}
+	offset := parseInt16(code[imm:])
+	dest := from + offset
+	if dest < 0 || dest >= length {
+		return NewParseError(ErrInvalidJumpDest, imm, "relative offset out-of-bounds: offset %d, dest %d", offset, dest)
+	}
+	if !analysis.codeSegment(uint64(dest)) {
+		return NewParseError(ErrInvalidJumpDest, imm, "relative offset into immediate value: offset %d, dest %d", offset, dest)
+	}
+	return nil
+}
+
+// validateControlFlow iterates through all possible branches the provided code
+// value and determines if it is valid per EOF v1.
+func validateControlFlow(code []byte, section int, metadata []*FunctionMetadata, jt *JumpTable) (int, error) {
+	type item struct {
+		pos    int
+		height int
+	}
+	var (
+		e              = NewParseError
+		heights        = make(map[int]int)
+		worklist       = []item{{0, int(metadata[section].Input)}}
+		maxStackHeight = int(metadata[section].Input)
+	)
+	for 0 < len(worklist) {
+		var (
+			idx    = len(worklist) - 1
+			pos    = worklist[idx].pos
+			height = worklist[idx].height
+		)
+		worklist = worklist[:idx]
+	outer:
+		for pos < len(code) {
+			op := OpCode(code[pos])
+
+			// Check if pos has already be visited; if so, the stack heights should be the same.
+			if want, ok := heights[pos]; ok {
+				if height != want {
+					return 0, e(ErrConflictingStack, pos, "have %d, want %d", height, want)
+				}
+				// Already visited this path and stack height
+				// matches.
+				break
+			}
+			heights[pos] = height
+
+			// Validate height for current op and update as needed.
+			if jt[op].minStack > height {
+				return 0, e(ErrStackUnderflow{stackLen: height, required: jt[op].minStack}, pos, "")
```

Comment:
You redefine `NewParseError` as `e` to make it look less crowded? Seems like a code-smell, to me. Perhaps there is some other way to simplify things. 
For example, why not just do this?

```
if have, want := jt[op].minStack, height; have > want{
    return 0, fmt.Errorf("%w: have %v want %v, pos %d", ErrStackUnderFlow, have, want, pos)
}
```

Reply:
Yeah that was a super wonky approach. I had an idea that it would be better to store some common metadata in the error to have a standard way of printing out, but just wrapping with `fmt.Errorf` seems much better.